design/0001-frr.md

```diff
@@ -0,0 +1,340 @@
+                              │               │  │               │
+                              └───────────────┘  └───────┬───────┘
+                                                         │
+                                                         │  <<use>>
+                                                 ┌───────▼───────┐
+                                                 │               │
+                                                 │    session    │
+                                                 │ <<interface>> │
+                                                 │               │
+                                                 └───────▲───────┘
+                                                         │ <<implement>>
+                                             ┌───────────┴─────────────┐
+                                             │                         │
+                                     ┌───────┴───────┐         ┌───────┴───────┐
+                                     │               │         │               │
+                                     │  bgp_metallb  │         │    bgp_frr    │
+                                     │  <<package>>  │         │  <<package>>  │
+                                     │               │         │               │
+                                     └───────────────┘         └───────┬───────┘
+                                                                       │
+                                                                       │
+                                                               ┌───────┴───────┐
+                                                               │               │
+                                                               │      frr      │
+                                                               │ <<container>> │
+                                                               │               │
+                                                               └───────────────┘
+
+```
+
+#### FRR deployment
+
+FRR will be deployed in a container as part of the speaker Pod. This will
+simplify the deployment for end users as MetalLB will not need to manage a
+another FRR Pod or deal with inter-Pod communication.
+
+Deploying FRR as a seperate Pod was also considered and it would give some
+advantages, such as the removal of the requirement for host networking from the
+speaker Pod, and the seperation of the FRR and speaker Pod lifecycles but
+would add complexity to the end-user.
+
+FRR can be controlled via a unix socket or a TCP socket. Connectivity via the
+unix socket is authorized through file permissions, whereas TCP connectivity is
+authorized through a password. FRR does not encyrpt traffic over the TCP socket
+which may be required at a later stage.
+
+Initially, control of the FRR container will be achieved by sending messages to
+localhost over the `vtysh` TCP socket, protected by a password. This will go
```

Comment:
This was discussed in a community call IIRC. The use case would be a deployment that already has an FRR instance deployed somewhere and wanted to reuse that instance (for example sharing it with another application). This is something we would like to do at Red Hat in the future. 

We can specify port and address to listen on 

```
       -A, --vty_addr vty-addr
              Specify an IP/IPv6 address to bind the TCP VTY interface to.  It is generally recommended to specify ::1 or 127.0.0.1.  For reasons of backwards compatibility, the default is to listen on all interfaces.
```

```
  -P, --vty_port vty-port
              Override the daemon's default TCP VTY port (each daemon has a different default value upwards of 2600, listed below.)  Specifying 0 disables the TCP VTY interface.

```

Reply:
> Can frr listen only on localhost? Just curious: why password protected? And do you envision such password as shared across all metallb downloads or unique per installation?
Just added security. Probably unique per node.

Reply:
(the reply quoted your answer too)

Thanks! Ok, if that is useful at Red Hat, seems better to use that on all cases rather than having a config knob just for that :)

Can you please mention the password will be unique per node? It is a detail I'd like to remember when reviewing the code, so better if it is explicit :)

Reply:
Done